services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/TerminalPowBlockMonitorService.java

```diff
@@ -0,0 +1,182 @@
+
+    asyncRunner
+        .runAfterDelay(this::monitor, Constants.TTD_MONITOR_SERVICE_POLL_INTERVAL)
+        .reportExceptions();
+  }
+
+  private void doMonitor() {
+    if (!isMerge.get()) {
+      initMergeState();
+      if (!isMerge.get()) {
+        return;
+      }
+    }
+
+    Optional<BeaconState> beaconState = recentChainData.getBestState();
+    if (beaconState.isEmpty()) {
+      return;
+    }
+
+    if (miscHelpers.isMergeComplete(beaconState.get())) {
+      stop();
+      return;
+    }
+
+    maybeBlockHashTracking.ifPresentOrElse(
+        this::checkTerminalBlockByBlockHash, this::checkTerminalBlockByTTD);
+  }
+
+  private void initMergeState() {
+    Optional<UInt64> maybeSlot = recentChainData.getCurrentSlot();
+    if (maybeSlot.isEmpty()) {
+      return;
+    }
+
+    Optional<SpecConfigMerge> maybeMergeSpecConfig =
+        spec.getSpecConfig(spec.computeEpochAtSlot(maybeSlot.get())).toVersionMerge();
+    if (maybeMergeSpecConfig.isEmpty()) {
+      return;
+    }
+
+    specConfigMerge = maybeMergeSpecConfig.get();
+    miscHelpers = ((MiscHelpersMerge) spec.atSlot(maybeSlot.get()).miscHelpers());
+
+    Optional<BeaconState> beaconState = recentChainData.getBestState();
+    if (beaconState.isEmpty()) {
+      return;
+    }
+
+    if (specConfigMerge.getTerminalBlockHash().isZero()) {
+      maybeBlockHashTracking = Optional.empty();
+    } else {
+      maybeBlockHashTracking = Optional.of(specConfigMerge.getTerminalBlockHash());
+    }
+
+    isMerge.set(true);
+  }
+
+  private void checkTerminalBlockByBlockHash(Bytes32 blockHashTracking) {
+    Optional<UInt64> slot = recentChainData.getCurrentSlot();
+    if (slot.isEmpty()) {
+      return;
+    }
+
+    final boolean isActivationEpochReached =
+        miscHelpers
+            .computeEpochAtSlot(slot.get())
+            .isGreaterThanOrEqualTo(specConfigMerge.getTerminalBlockHashActivationEpoch());
+
+    if (isActivationEpochReached) {
+      executionEngine
+          .getPowBlock(blockHashTracking)
+          .thenAccept(
+              maybePowBlock ->
+                  maybePowBlock
+                      .map(PowBlock::getBlockHash)
+                      .map(blockHashTracking::equals)
+                      .ifPresent(
+                          found -> {
+                            if (found) {
+                              onTerminalPowBlockFound(blockHashTracking);
+                            }
+                          }))
+          .finish(
+              error -> LOG.error("Unexpected error while searching Terminal Block by Hash", error));
+    }
+  }
+
+  private void checkTerminalBlockByTTD() {
+    executionEngine
+        .getPowChainHead()
+        .thenAccept(
+            powBlock -> {
+              UInt256 difficulty = powBlock.getTotalDifficulty();
+              if (difficulty.compareTo(specConfigMerge.getTerminalTotalDifficulty()) < 0) {
+                onTerminalPowBlockFound(powBlock.getBlockHash());
+              }
```

Comment:
Wouldn't the terminal block need to have a total difficulty greater than or equal to the terminal total difficulty? The parent of the terminal block has to be less than TTD, but the terminal block itself must be equal to or greater.

Reply:
🤦